File: archive/pulumi-py/Pr4316/lib/rds_stack.py
"""
RDS Stack - Aurora Serverless PostgreSQL database.

This module creates an encrypted Aurora Serverless v2 PostgreSQL cluster
for storing payment transaction data with PCI-DSS compliant configuration.
"""

from typing import Optional, List
import pulumi
import pulumi_aws as aws
from pulumi import ResourceOptions


class RdsStackArgs:
    """
    Arguments for RDS Stack.

    Args:
        environment_suffix: Environment identifier for resource naming
        subnet_ids: List of private subnet IDs for RDS
        security_group_id: Security group ID for RDS
        kms_key_arn: KMS key ARN for encryption
    """
    def __init__(
        self,
        environment_suffix: str,
        subnet_ids: List[pulumi.Output],
        security_group_id: pulumi.Output,
        kms_key_arn: pulumi.Output
    ):
        self.environment_suffix = environment_suffix
        self.subnet_ids = subnet_ids
        self.security_group_id = security_group_id
        self.kms_key_arn = kms_key_arn


class RdsStack(pulumi.ComponentResource):
    """
    RDS Component Resource for transaction database.

    Creates Aurora Serverless v2 PostgreSQL cluster with:
    - Encryption at rest using KMS
    - Automated backups
    - Private subnet placement
    - PCI-DSS compliant configuration
    """

    def __init__(
        self,
        name: str,
        args: RdsStackArgs,
        opts: Optional[ResourceOptions] = None
    ):
        super().__init__('custom:database:RdsStack', name, None, opts)

        # DB Subnet Group for RDS placement in private subnets
        self.db_subnet_group = aws.rds.SubnetGroup(
            f"payment-db-subnet-group-{args.environment_suffix}",
            name=f"payment-db-subnet-group-{args.environment_suffix}",
            subnet_ids=args.subnet_ids,
            tags={
                "Name": f"payment-db-subnet-group-{args.environment_suffix}",
                "Environment": args.environment_suffix,
            },
            opts=ResourceOptions(parent=self)
        )

        # PCI-DSS Requirement: Encrypted storage for cardholder data
        # Aurora Serverless v2 for cost optimization and faster provisioning
        self.db_cluster = aws.rds.Cluster(
            f"payment-db-cluster-{args.environment_suffix}",
            cluster_identifier=f"payment-db-cluster-{args.environment_suffix}",
            engine=aws.rds.EngineType.AURORA_POSTGRESQL,
            engine_mode="provisioned",  # Required for Serverless v2
            engine_version="15.5",
            database_name="paymentdb",
            master_username="dbadmin",
            master_password="ChangeMe123!",  # In production, use AWS Secrets Manager
            db_subnet_group_name=self.db_subnet_group.name,
            vpc_security_group_ids=[args.security_group_id],
            # PCI-DSS Requirement: Encryption at rest
            storage_encrypted=True,
            kms_key_id=args.kms_key_arn,
            # Backup configuration for data protection
            backup_retention_period=1,  # Minimum for faster creation
            preferred_backup_window="03:00-04:00",
            # Disable deletion protection for test environments
            deletion_protection=False,
            skip_final_snapshot=True,
            # Serverless v2 scaling configuration
            serverlessv2_scaling_configuration=aws.rds.ClusterServerlessv2ScalingConfigurationArgs(
                min_capacity=0.5,
                max_capacity=2.0,
            ),
            tags={
                "Name": f"payment-db-cluster-{args.environment_suffix}",
                "Environment": args.environment_suffix,
                "Compliance": "PCI-DSS",
            },
            opts=ResourceOptions(parent=self)
        )

        # Aurora Serverless v2 instance
        self.db_instance = aws.rds.ClusterInstance(
            f"payment-db-instance-{args.environment_suffix}",
            identifier=f"payment-db-instance-{args.environment_suffix}",
            cluster_identifier=self.db_cluster.id,
            instance_class="db.serverless",
            engine=self.db_cluster.engine,
            engine_version=self.db_cluster.engine_version,
            tags={
                "Name": f"payment-db-instance-{args.environment_suffix}",
                "Environment": args.environment_suffix,
            },
            opts=ResourceOptions(parent=self, depends_on=[self.db_cluster])
        )

        # Register outputs
        self.register_outputs({
            "db_cluster_endpoint": self.db_cluster.endpoint,
            "db_cluster_reader_endpoint": self.db_cluster.reader_endpoint,
            "db_cluster_arn": self.db_cluster.arn,
            "db_name": self.db_cluster.database_name,
        })

File: archive/cdktf-ts/Pr7584/lib/rds-stack.ts
import { RdsCluster } from '@cdktf/provider-aws/lib/rds-cluster';
import { RdsClusterInstance } from '@cdktf/provider-aws/lib/rds-cluster-instance';
import { DbSubnetGroup } from '@cdktf/provider-aws/lib/db-subnet-group';
import { SecurityGroup } from '@cdktf/provider-aws/lib/security-group';
import { Subnet } from '@cdktf/provider-aws/lib/subnet';
import { KmsKey } from '@cdktf/provider-aws/lib/kms-key';
import { SecretsmanagerSecret } from '@cdktf/provider-aws/lib/secretsmanager-secret';
import { Construct } from 'constructs';

export interface RdsStackProps {
  environmentSuffix: string;
  privateSubnet1: Subnet;
  privateSubnet2: Subnet;
  securityGroup: SecurityGroup;
  kmsKey: KmsKey;
  dbSecret: SecretsmanagerSecret;
  dbUsername: string;
  dbPassword: string;
}

export class RdsStack extends Construct {
  public readonly cluster: RdsCluster;

  constructor(scope: Construct, id: string, props: RdsStackProps) {
    super(scope, id);

    const {
      environmentSuffix,
      privateSubnet1,
      privateSubnet2,
      securityGroup,
      kmsKey,
      dbUsername,
      dbPassword,
    } = props;

    // Create DB Subnet Group
    const subnetGroup = new DbSubnetGroup(this, 'db-subnet-group', {
      name: `assessment-db-subnet-${environmentSuffix}`,
      description: 'Subnet group for assessment RDS cluster',
      subnetIds: [privateSubnet1.id, privateSubnet2.id],
      tags: {
        Name: `assessment-db-subnet-${environmentSuffix}`,
      },
    });

    // Create Aurora Serverless v2 Cluster (PostgreSQL)
    this.cluster = new RdsCluster(this, 'aurora-cluster', {
      clusterIdentifier: `assessment-cluster-${environmentSuffix}`,
      engine: 'aurora-postgresql',
      engineMode: 'provisioned',
      engineVersion: '15.4',
      databaseName: 'assessmentdb',
      masterUsername: dbUsername,
      masterPassword: dbPassword,
      dbSubnetGroupName: subnetGroup.name,
      vpcSecurityGroupIds: [securityGroup.id],
      storageEncrypted: true,
      kmsKeyId: kmsKey.arn,
      backupRetentionPeriod: 7,
      preferredBackupWindow: '03:00-04:00',
      preferredMaintenanceWindow: 'sun:04:00-sun:05:00',
      skipFinalSnapshot: true,
      deletionProtection: false,
      enabledCloudwatchLogsExports: ['postgresql'],
      serverlessv2ScalingConfiguration: {
        minCapacity: 0.5,
        maxCapacity: 1.0,
      },
      tags: {
        Name: `assessment-cluster-${environmentSuffix}`,
      },
    });

    // Create Aurora Serverless v2 Instance (PostgreSQL)
    new RdsClusterInstance(this, 'aurora-instance-1', {
      identifier: `assessment-instance-1-${environmentSuffix}`,
      clusterIdentifier: this.cluster.id,
      instanceClass: 'db.serverless',
      engine: 'aurora-postgresql',
      engineVersion: '15.4',
      tags: {
        Name: `assessment-instance-1-${environmentSuffix}`,
      },
    });
  }
}
